support defaultValues for query variables

srtucker22 opened the pull request:
<!--
  Thanks for filing a pull request on Apollo Client!

  Please look at the following checklist to ensure that your PR
  can be accepted quickly:
-->

TODO:

- [x] If this PR is a new feature, reference an issue where a consensus about the design was reached (not necessary for small changes)
- [x] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass
- [x] Update CHANGELOG.md with your change
- [x] Add your name and email to the AUTHORS file (optional)
- [x] If this was a change that affects the external API, update the docs and post a link to the PR in the discussion

srtucker22 commented:
Hi there,

Sorry for not getting back to you on this -- I've been traveling for the last couple months and finally got a chance to look at it today. Not sure where the best place is to put tests, but I made two basic ones in `test/client.ts`

srtucker22 commented:
Also noticed `graphql` didn't seem to flag queries with default values for non-null variables as [per the spec](http://facebook.github.io/graphql/#sec-Variable-Default-Values-Are-Correctly-Typed)? Maybe this sort of logic belongs in `getFromAST` or a change to `graphql` package?

stubailo commented:
@srtucker22 `gql` intentionally doesn't validate queries at all. That's the job of the graphql server, or a tool like eslint-plugin-graphql: https://github.com/apollographql/eslint-plugin-graphql

stubailo commented on src/core/QueryManager.ts:
```diff
@@ -482,8 +482,19 @@ export class QueryManager {
+    if (queryDefinition.variableDefinitions && queryDefinition.variableDefinitions.length) {
+      const defaultValues = queryDefinition.variableDefinitions
+                              .filter(({ defaultValue }) => defaultValue)
+                              .map(({ variable, defaultValue }) => ({ [variable.name.value]: defaultValue }));
```
Will this work for more complex default values such as:

```graphql
4
"string"
{ key: "value" }
["array1", "array2"]
{ key: ["array1", "array2"] }
```

I think we can reuse the method from here: https://github.com/apollographql/apollo-client/blob/dfb12ea39da749a26f46767b43993eb08ba4d66a/src/data/storeUtils.ts#L50

That's a function that knows how to accept a GraphQL AST value and product a JavaScript object.

stubailo commented on src/core/QueryManager.ts:
```diff
@@ -508,7 +511,21 @@ export class QueryManager {
+                                  return;
+                                }
+
+                                const v: any = {};
```
Can we use a variable name that's not `v` here? And not use the `any` type?

stubailo commented on src/core/QueryManager.ts:
```diff
@@ -508,7 +511,21 @@ export class QueryManager {
       const defaultValues = queryDefinition.variableDefinitions
                               .filter(({ defaultValue }) => defaultValue)
-                              .map(({ variable, defaultValue }) => ({ [variable.name.value]: defaultValue }));
+                              .map(({ variable, defaultValue }) : any => {
```
No `any` type please, let's import the correct `ValueNode` or something from `graphql`.

stubailo commented on src/core/QueryManager.ts:
```diff
@@ -501,8 +504,33 @@ export class QueryManager {
+                                  defaultValue,
+                                );
+
+                                return {[`$${variable.name.value}`]: v[variable.name.value]};
```
This is not correct - variables in the dictionary sent to the server don't have `$` in the name.

srtucker22 replied:
whoops!

stubailo commented on test/client.ts:
```diff
@@ -267,6 +267,82 @@ describe('client', () => {
+      },
+    };
+
+    clientRoundrip(query, data);
```
I don't think this test is actually testing the variables at all - you need a test that actually specifies what is sent to the server, because I don't think the implementation here is working. In fact, this is confirming that the code doesn't work I think, because if you look at the `clientRoundtrip` implementation it doesn't actually include any variables.

I think you need to write some custom code (like what `clientRoundtrip` has inside) to specify the exact requests and responses you are expecting.

srtucker22 replied:
yep

stubailo commented on test/client.ts:
```diff
@@ -267,6 +267,82 @@ describe('client', () => {
+      },
+    };
+
+    const variables = { $first: 1 };
```
The correct variables here would be `{ first: 1 }`

srtucker22 commented:
so the previous commit actually didn't work at all -- thought it did because the tests were green but i actually don't see `clientRoundrip` (`clientRoundtrip`?) ever reaching `assert.deepEqual(result.data, data);` for any tests in `client.ts`

new commit didn't fix the tests but the output from debugging looks correct including complex inputs thanks to the utils function you suggested. surely, there's a more elegant way to write the code, but wanted to get something in now that's at least working before going off the grid again :)  just landed from a redeye in Osaka and won't be online for a few days at least.

stubailo commented on src/core/QueryManager.ts:
```diff
@@ -624,8 +627,30 @@ export class QueryManager {
+                              .filter(({ defaultValue }) => defaultValue)
+                              .map(({ variable, defaultValue }) : { [key: string]: any } => {
+                                if (!defaultValue) {
+                                  return {}; // should never reach
```
Any point in having this condition here if it can never happen? I don't think we need to guard against incorrect ASTs since the parser would have failed already.

stubailo commented on src/core/QueryManager.ts:
```diff
@@ -624,8 +627,30 @@ export class QueryManager {
+    // assign variable default values if supplied
+    if (queryDefinition.variableDefinitions && queryDefinition.variableDefinitions.length) {
+      const defaultValues = queryDefinition.variableDefinitions
+                              .filter(({ defaultValue }) => defaultValue)
```
I'd prefer this to be indented two spaces only.

srtucker22 commented:
great! i'll look into factoring it out for mutations/subscriptions when i get a sec

srtucker22 commented:
hey, the latest changes should have default vals working properly for mutations and subscriptions, i think

let me know what's good 🍺

helfer commented:
Hey @srtucker22, really awesome work here! 👍 

There's one more thing we forgot here, which is making sure that it works with the new `client.readQuery`, `client.writeQuery`, `client.readFragment` functions as well as the `update` functions. Can you add the necessary code and tests for that? I'll fast-track this PR so it'll be reviewed and merged as soon as you're done coding! ⏩

srtucker22 commented:
Hey @helfer,

I've added some code that I think should cover those functions. But I do have a couple questions:
1. how do you declare variables for just a fragment -- for `client.readFragment`?
2. not sure how/where you're testing `update` functions, but I've added default values to `diffQueryAgainstStore` and `writeResultToStore` which I think hits all the update funcs, but I could be very wrong

Let me know what's good

helfer commented:
@srtucker22 I think for fragments there's currently no way to define variables. I think that's okay, we can just leave it as is.

`diffQueryAgainstStore` and `writeResultToStore` are indeed the two functions that will cover everything, so I'm pretty optimistic it'll be fine.

I don't have time to review right now, but I'll do it tomorrow!

helfer commented on test/ApolloClient.ts:
```diff
@@ -103,6 +103,41 @@ describe('ApolloClient', () => {
 
+  it('will read some data from the store with default values', () => {
+      const client = new ApolloClient({
+        initialState: {
```
A neat trick here would be to use `client.writeQuery`.
